sitespeedio/sitespeed.io
# fix(plugins/slack): correctly send messages on html.finished message

### Checklist for a pull request to sitespeed.io

- [x] I'm making a big change or adding functionality so I've already opened an issue proposing the change to other contributors, so I got feedback on the idea before took the time to write precious code
- [x] Check that your change/fix has corresponding unit tests (if applicable)
- [x] Squash commits so it looks sane
- [x] ~~Update the documentation https://github.com/sitespeedio/sitespeed.io/tree/master/docs in another PR~~
- [x] Verify that the test works by running `npm test` and test linting by `npm run lint`

### Description

Do not rely on `resultBaseUrl` to block messages to be sent to Slack. Instead, check if any
of the supported hosting providers are correctly configured. If that is the case, then
messages are sent _only after the upload is complete_. For all other cases, send the message
anyway.

**Add basic test scenarios**:  

- Check for missing options when calling `slack.open()`
- Ensure that methods supposed to return lists always return them
- Add some sanity checks to the summary, so that missing values would not crash

**Notes**:

Most test cases were included because I had to wrap my head around the whole Sitespeed architecture and how the `postMessage()` calls from the queue would end up in the Slack plugin. While doing it so, I noticed that when calling the plugin with some specific test data (i.e, missing arguments, or `undefined`) the plugin would crash. I do expect these cases to not happen when running sitespeed but I took the approach of _better safe than sorry_ here. (i.e 
 a `getURLs()` method returning `undefined` doesn't look correct as I would expect it to _always_ return a list, even if it's an empty one). 

---

Closes #2147
